src/s3ql/backends/b2/b2_backend.py

```diff
@@ -0,0 +1,768 @@
+
+        try:
+            response, body = self._do_request(self._get_download_connection(), method, path, download_body=False)
+        except HTTPError as exc:
+            if exc.status == 404:
+                raise NoSuchObject(key)
+
+        return response
+
+    def _do_upload_request(self, headers=None, body=None):
+        upload_url_info = self._get_upload_url_info()
+        headers['Authorization'] = upload_url_info['authorizationToken']
+
+        upload_url_info['isUploading'] = True
+
+        try:
+            response, response_body = self._do_request(upload_url_info['connection'], 'POST', upload_url_info['path'], headers, body)
+        except B2Error as exc:
+            if exc.status == 503:
+                # storage url too busy, change it
+                self._invalidate_upload_url(upload_url_info)
+
+            raise
+
+        except (ConnectionClosed, StateError):
+            # storage url too busy, change it
+            self._invalidate_upload_url(upload_url_info)
+            raise
+
+        except:
+            upload_url_info['isUploading'] = False
+            raise
+
+        upload_url_info['isUploading'] = False
+
+        json_response = json.loads(response_body.decode('utf-8'))
+
+        log.debug('UPLOAD RESPONSE:', json.dumps(json_response, indent=2))
+
+        return json_response
+
+    def _get_bucket_id(self):
+        if not self.bucket_id:
+            request_data = {
+                'accountId': self._get_account_id(),
+                'bucketName': self.bucket_name
+            }
+            response = self._do_api_call('b2_list_buckets', request_data)
+
+            buckets = response['buckets']
+            for bucket in buckets:
+                if bucket['bucketName'] == self.bucket_name:
+                    self.bucket_id = bucket['bucketId']
+
+            if not self.bucket_id:
+                raise DanglingStorageURLError(self.bucket_name)
+
+        return self.bucket_id
+
+    @property
+    @copy_ancestor_docstring
+    def has_native_rename(self):
+        return False
+
+    @property
+    @copy_ancestor_docstring
+    def has_delete_multi(self):
+        return False
+
+    @copy_ancestor_docstring
+    def is_temp_failure(self, exc):
+        if is_temp_network_error(exc) or isinstance(exc, ssl.SSLError):
+            # We better reset our connections
+            self._reset_connections()
+
+        if is_temp_network_error(exc):
+            return True
+
+        elif (isinstance(exc, B2Error) and
+              (exc.code == 'bad_auth_token' or
+               exc.code == 'expired_auth_token')):
+            return True
+
+        elif (isinstance(exc, B2Error) and
+              ((500 <= exc.status <= 599) or
+               exc.status == 408 or
+               exc.status == 429)):
+            return True
+
+        elif (isinstance(exc, HTTPError) and
+              exc.status == 408):
+            return True
+
+        elif isinstance(exc, StateError):
```

Comment:
I am not sure about this, I got a `StateError` occasionally and this seemed to fix it.

Reply:
If you get a `StateError`, this means there is a bug in how you're using dugong, or in dugong itself. In either case, the above is a workaround that hides the real problem. So no, this isn't the a good way to do it.

Reply:
That's what I was thinking. I had a short look into the dugong documentation (although due to lack of time not very extensively I admit), but it didn't completely clear up the way it works, so I'll dig into that a bit more. Do you have by chance some hints how to log/debug this?